extra/dist/world.py

```diff
@@ -1,59 +1,101 @@
-from typing import Any, Optional, Tuple
+from typing import Optional
 from extra import dist
 from multiprocessing import shared_memory
-from tinygrad.helpers import DEBUG, colored
+from tinygrad.helpers import DEBUG, colored, getenv
 from tinygrad.lazy import LazyBuffer
-from tinygrad.runtime.lib import RawBufferCopyIn, RawBufferCopyInOut
+from tinygrad.runtime.lib import RawBuffer, RawBufferCopyIn, RawBufferCopyInOut
+try: from tinygrad.runtime.ops_hip import RawHIPBuffer
+except: RawHIPBuffer = None
 from tinygrad.runtime.ops_shm import RawShmBuffer
 from tinygrad.jit import CacheCollector
 from tinygrad.tensor import Tensor, Function
+import extra.hip_wrapper as hip
 import numpy as np
 
 # fake the function signature of ASTRunner so we can put it in the cache
-def __send_rb(args:Tuple[RawBufferCopyInOut, RawShmBuffer, int, Any], variables=None, jit=False, force_wait=False):
-  args[0]._copyout(np.frombuffer(args[1]._buffer(), dtype=args[0].dtype.np))
-  dist.OOB.send(args[3], args[2])
-  if DEBUG >= 2: print(f"{colored('****', 'magenta' if jit else None)}   sent {args[0]} to rank {args[2]}")
+def __send_rb(args, variables=None, jit=False, force_wait=False):
+  x, target_rank, y = args[:3]
+  if RawHIPBuffer and x.__class__ is RawHIPBuffer:
+    hip.hipSetDevice(x._device)
+    hip.hipDeviceSynchronize()
```

Comment:
Hmm. So this is okay, it's better than what we have, but is there a way to do this without needing to sync the GPU?